### PR TITLE
Feature arduino gps

### PR DESCRIPTION
**Pull request by itzandroidtab**
This removes most of the calculations with floats. And this makes location detector work with a arduino.

### CODE REVIEW

File: code/headers/gga.hpp [hunk 1]
```diff
@@ -3,6 +3,17 @@
 #include <hwlib.hpp>
 
 namespace r2d2::location {
+    /**
+     * @brief Struct containing the degrees, minutes, tenthousends minute for
+     * latitude/longitude.
+     *
+     */
+    struct degrees {
+        uint16_t tenthousends;
```
Comment (florianhumblot) [inline code/headers/gga.hpp]: Should be `tenthousandths`

File: code/headers/gga.hpp [hunk 2]
```diff
@@ -3,6 +3,17 @@
 #include <hwlib.hpp>
 
 namespace r2d2::location {
+    /**
+     * @brief Struct containing the degrees, minutes, tenthousends minute for
+     * latitude/longitude.
+     *
+     */
+    struct degrees {
```
Comment (florianhumblot) [inline code/headers/gga.hpp]: Is it really a good idea to call the structs degrees when it has a member of the same name?
It also isn't really descriptive.

File: code/headers/gga.hpp [hunk 3]
```diff
@@ -11,18 +22,81 @@ namespace r2d2::location {
...
+        // 0 = invalid
+        // 1 = GPS fix (SPS)
+        // 2 = DGPS fix, OmniSTAR VBS
+        // 3 = PPS fix
+        // 4 = Real Time Kinematic, fixed integers
+        // 5 = Float RTK, float integers, OmniSTAR XP/HP or Location RTK
+        // 6 = estimated (dead reckoning) (2.3 feature)
+        // 7 = Manual input mode
+        // 8 = Simulation mode
+        uint8_t fix_quality;
```
Comment (florianhumblot) [inline code/headers/gga.hpp]: Wouldn't this be the perfect case for an `enum class`?
considering you have 8 distinct values with a meaning.

File: code/src/nmea_parser.cpp [hunk 4]
```diff
@@ -29,19 +60,19 @@ namespace r2d2::location {
                 break;
             }
             case GGA::latitude: {
-                location.latitude = atof(gps_message + i, sep_offset);
+                location.latitude = get_degrees(gps_message + i, sep_offset);
                 break;
             }
             case GGA::north_south_hemisphere: {
-                location.north_south_hemisphere = gps_message[i];
+                location.is_north_hemisphere = gps_message[i] == 'N';
```
Comment (florianhumblot) [inline code/src/nmea_parser.cpp:67]: Why is a `char` being put in a `bool` member?

File: code/src/nmea_parser.cpp [hunk 5]
```diff
@@ -29,19 +60,19 @@ namespace r2d2::location {
...
                 break;
             }
             case GGA::longitude: {
-                location.longitude = atof(gps_message + i, sep_offset);
+                location.longitude = get_degrees(gps_message + i, sep_offset);
                 break;
             }
             case GGA::east_west_hemisphere: {
-                location.east_west_hemisphere = gps_message[i];
+                location.is_east_hemisphere = gps_message[i] == 'E';
```
Comment (florianhumblot) [inline code/src/nmea_parser.cpp:75]: Why is a `char` being put in a `bool` member?

File: code/src/nmea_parser.cpp [hunk 6]
```diff
@@ -61,15 +92,35 @@ namespace r2d2::location {
...
                 location.geoid_height = atof(gps_message + i, sep_offset);
                 break;
             }
             case GGA::geoid_height_measurement: {
-                location.geoid_height_measurement = gps_message[i];
+                location.geoid_height_unit = gps_message[i];
+                break;
+            }
+            case GGA::diff_age: {
+                // this parameters is not transmitted if we dont have a dgps fix
```
Comment (florianhumblot) [inline code/src/nmea_parser.cpp]: Parameter should be singular

File: code/src/nmea_parser.cpp [hunk 7]
```diff
@@ -61,15 +92,35 @@ namespace r2d2::location {
...
+                // or higher.
```
Comment (florianhumblot) [inline code/src/nmea_parser.cpp]: This doesn't really make sense, what does 'higher than a dgps fix' mean?

File: code/src/nmea_parser.cpp [hunk 8]
```diff
@@ -61,15 +92,35 @@ namespace r2d2::location {
...
+                if (location.fix_quality <= 1) {
+                    break;
+                }
+
+                location.diff_age = atoi(gps_message + i, sep_offset);
+                break;
+            }
+            case GGA::diff_id: {
+                // this parameters is not transmitted if we dont have a dgps fix
```
Comment (florianhumblot) [inline code/src/nmea_parser.cpp]: Parameter should be singular

**Review by florianhumblot — CHANGES_REQUESTED** [review]
A few grammar nitpicks as well as some questionable design decisions.
please either explain the Rationale behind some of these decisions or change them.

File: code/src/nmea_parser.cpp (hunk 4 above)
Comment (itzandroidtab) [reply code/src/nmea_parser.cpp:67]: There is no `char` in a bool. As `gps_message[i] == 'N'` returns `true` or `false`

File: code/src/nmea_parser.cpp (hunk 5 above)
Comment (itzandroidtab) [reply code/src/nmea_parser.cpp:75]: Same as previous message

File: code/src/nmea_parser.cpp (hunk 7 above)
Comment (itzandroidtab) [reply code/src/nmea_parser.cpp]: This refers to the values in the `gga_s` struct. Higher than dgps =  `>= 2`. I have added some explanation to the comment as well

File: code/headers/gga.hpp (hunk 2 above)
Comment (itzandroidtab) [reply code/headers/gga.hpp]: changed the name to `decimal_degrees`.

https://en.wikipedia.org/wiki/Decimal_degrees

File: code/src/nmea_parser.cpp (hunk 4 above)
Comment (florianhumblot) [reply code/src/nmea_parser.cpp:67]: My  bad, I misread the statement.

File: code/src/nmea_parser.cpp (hunk 5 above)
Comment (florianhumblot) [reply code/src/nmea_parser.cpp:75]: My bad, misread the statement.

**Review by florianhumblot — APPROVED** [review]

**Pull request merged**